server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -590,6 +598,11 @@ void executeBulk(
         Executor executor,
         AtomicArray<BulkItemResponse> responses
     ) {
+        // Determine if we have the feature enabled once for entire bulk operation
+        final boolean clusterSupportsFailureStore = featureService.clusterHasFeature(
```

Comment:
I assume even for a very large cluster, this check is cheap enough relatively to a bulk request that it's fine to run on every bulk? I think it just loops through all the nodes in the cluster state.

Reply:
It is indeed a linear check over the list of nodes in the cluster. Unfortunately there's not a great way to hoist this up further than a request-by-request basis without introducing some kind of timer element or observer interface. We need to be responsive to changes in the cluster in a timely manner, and refactoring things to build off a cluster state listener seems like more complexity than is worth at the moment. I'd be hard pressed to try and optimize this further without indication that it's in a bad place.